keps/0012-framework-extensions.md

```diff
@@ -0,0 +1,319 @@
+    - [Story 2](#story-2)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints-optional)
+  - [Outstanding Questions](#outstanding-questions)
+- [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+[tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+Generic implementations of frameworks don't capture all customizations that are required for all end users. Rather than having to fork and maintain a patched version of a Framework, Extending a framework will allow for additions and customizations to be made to frameworks to meet additional business needs.
+
+## Motivation
+
+Extensions and bases describe a mechanism for building frameworks or extensions to frameworks from a given base. In this document, an **extension** is any KUDO framework that extends from some base. A **base** is the complete set of manifests, metadata, and other files provided by that base's type. A base should provide complete information that users of that base tool are expected to have. The base types and what they expose to frameworks that extend from them are described in their respective sub-sections.
+
+### Goals
+
+- Be able to extend an existing Framework
+
+### Non-Goals
+
+- Extending Helm Charts, CNAB or operators
+- Managing repos of extensions
+
+## Proposal
+
+### User Stories
+
+- As an Application Operator, I want to be able to add a custom Plan to an existing FrameworkVersion.
+- As an Application Operator, I want to be able to patch existing templates in an existing FrameworkVersion.
+- As an Application Operator, I want to be able to add or update parameters to an existing FrameworkVersion.
+- As an Application Operator, I want to be able to update my base framework easily.
+
+### Extension Spec
+
+To support extending from a base, `framework.yaml` is extended to support the `extends` keyword:
+
+```yaml
+extends:
+  kudo:
+    framework: "mysql"
+    version: "5.7"
+```
+
+What gets inherited? Everything. A framework defined as:
+
+```yaml
+framework: bar
+extends:
+  kudo:
+    framework: "foo"
+    version: "1.0
+```
+
+will have the EXACT same functionality (plans, parameters, tasks, templates, etc) as the base `foo` framework.
+
+#### Refrencing an object from the base
+
+#### Adding a Task
+
+Tasks can be added via the same `tasks` spec in the `Framework` definition. Templates used to define the task can come from the base framework by pre-fixing `base/` to the template name, or from local templates as defined in [KEP-0009](keps/0009-operator-toolkit.md). The follow example shows a new task called `load-data` defined for the extension framework that uses both a template from the base, with a patch object that is defined in the extention framework.
+
+```yaml
+extends:
+  kudo:
+    framework: "mysql"
+    version: "5.7"
+version: "5.7"
+tasks:
+  load-data:
+    resources:
+      - base/init.yaml
+    patches:
+      - load-data.yaml
```

Comment:
Could you please specify how does `patches` exactly work? Is it strategic merge patch?

Reply:
same here

Reply:
https://github.com/kudobuilder/kudo/issues/55